src/platform/screen.c

```diff
@@ -1,22 +1,141 @@
 #include "platform/screen.h"
-
-#include "SDL.h"
+#include "platform/dpi.h"
 
 #include "game/settings.h"
 #include "graphics/graphics.h"
 #include "graphics/screen.h"
 
+#include "SDL.h"
+#include <math.h>
+#include <stdlib.h>
+
 static struct {
     SDL_Window *window;
     SDL_Renderer *renderer;
     SDL_Texture *texture;
-} SDL = {0, 0, 0};
+} SDL;
 
 static struct {
     int x;
     int y;
+    int w;
+    int h;
+    int moving;
 } window_pos;
 
+static struct {
+    const int WIDTH;
+    const int HEIGHT;
+} MINIMUM = { 640, 480 };
+
+/* Since resize events are called from a separate thread, we need to prevent unwanted resizes. */
+static int prevent_threaded_resize_event = 1;
+
+static void platform_screen_position_window(int width, int height)
+{
+    int title_height = platform_dpi_get_title_height();
+
+    /* No position information, center window */
+    if ((!window_pos.x && !window_pos.y) || !window_pos.w || !window_pos.h) {
+        int display = SDL_GetWindowDisplayIndex(SDL.window);
+        SDL_SetWindowPosition(SDL.window, SDL_WINDOWPOS_CENTERED_DISPLAY(display), SDL_WINDOWPOS_CENTERED_DISPLAY(display));
+        SDL_GetWindowPosition(SDL.window, &window_pos.x, &window_pos.y);
+    } else if (width != window_pos.w || height != window_pos.h) {
+        /* New window size, use the window center as a reference */
+        window_pos.x = window_pos.x + (window_pos.w / 2) - (width / 2);
+        if (!window_pos.moving) {
+            /* If we got a reposition request while moving the window, it was because the window was changed to another monitor with a different DPI.
+               Keep the window title bar at the same position in that case. */
+            window_pos.y = window_pos.y + (window_pos.h / 2) - (height / 2);
+        }
+    }
+
+    SDL_Rect monitor_coordinates;
+    SDL_GetDisplayBounds(SDL_GetWindowDisplayIndex(SDL.window), &monitor_coordinates);
+    int minimum_y = monitor_coordinates.y + title_height;
+    int minimum_x = 0;
+    int total_displays = SDL_GetNumVideoDisplays();
+    for (int display_index = 0; display_index < total_displays; display_index++) {
+        SDL_GetDisplayBounds(display_index, &monitor_coordinates);
+        minimum_x = (monitor_coordinates.x < minimum_x) ? monitor_coordinates.x : minimum_x;
+    }
+
+    window_pos.x = (window_pos.x < minimum_x) ? minimum_x : window_pos.x;
```

Comment:
For my own reference: Should there also be a check for `maximum_x` and `maximum_y`? I guess it wouldn't hurt...